Improve logger interaction with Geant4

sethrj opened the pull request:
Similar to #773, this refactors the Geant4 `G4coutDestination` redirection into an API-level class that can be enabled with or  without Geant4 linkage. It also improves redirected output in two ways:
- Individual "redirected" lines are sent separately to the logger
- Logger messages are sent to `std::clog` rather than `std::cerr` so that they go to the `STDERR` stream but are buffered and don't get redirected when `std::err`'s destination is changed.

sethrj committed:
Add the ability to disable Geant4 signal handlers

sethrj committed:
Improve G4GDML logging

sethrj committed:
Use std::clog for Logger for buffering and to avoid redirection issues

sethrj committed:
Refactor Geant logger adapter into ScopedGeantLogger

This class is now thread-safe, allows multiple nested instances, and
fixes the problems with log messages being sent into another scoped IO
helper.

sethrj committed:
Add timer and improve documentation

pcanal commented on src/celeritas/ext/GeantSetup.cc:
```diff
@@ -138,6 +138,8 @@ GeantSetup::GeantSetup(std::string const& gdml_filename, Options options)
 
         // Construct the physics
+        CELER_LOG(status) << "Initializing Geant4 phyics";
+        ScopedTimeLog scoped_time;
```
It is not obvious (from just reading the diff) why here the `ScopedTimeLog` is used 'late', while in other case it is used _before_ `ScopedGeantLogger`.

sethrj replied:
I don't think it really matters since the end of scope is what counts most (i.e. it prints there) and the log statement takes effectively zero time.  I think it makes more sense to put it below the status? I think this particular instance should be removed since constructing the physics list is so cheap (it's building the tables that can be expensive).

pcanal commented on src/celeritas/ext/ScopedGeantLogger.cc:
```diff
@@ -0,0 +1,159 @@
+
+        if (!g_adapter_active_)
+        {
+            g_adapter_active_ = true;
+            logger_ = std::make_unique<GeantLoggerAdapter>();
```
```suggestion
            logger_ = std::make_unique<GeantLoggerAdapter>();
            g_adapter_active_ = true;
```
With the previous order, you can have another thread that go through this routine with `g_adapter_active_ == true` and `logger_ == nullptr` (or worse, since `logger_` is not an atomic, it "could" be a partially updated value) and thus giving the incorrect behavior (including possibly a crash)

pcanal replied:
Never mind.  `ScopedGeantLogger` is intended to be created on the stack and thus `ScopedGeantLogger::logger_` is not a shared resources.   The code seems correct to implement that "extra" `ScopedGeantLogger` will be ignored.

pcanal commented on src/celeritas/ext/ScopedGeantLogger.cc:
```diff
@@ -0,0 +1,159 @@
+{
+    if (logger_)
+    {
+        logger_.reset();
```
Do we need to take the `capture_mutex` lock here?

pcanal replied:
Not here but below.  Technically `g_adapter_active_` needs to either be an atomic or its update ought to be covered by the mutex.  In practice it may have no important visible effect (i.e. at worse another thread may not be able to install a the logger even-though it could of but it would also be the case with small variation in the timing of the threads).

If we indeed, choose to leave as-is, we ought to add some comment in the code explaining why.

sethrj replied:
I realized after a little more digging that the G4 output elements are thread-local: so we don't need mutexes at all, but the `g_adapter_active_` needs to be thread local. I've adjusted accordingly.

pcanal replied:
Which commit implement this adjustment?

sethrj replied:
Uhhhh apparently one that I didn't actually `git commit` 😞 this is why we have reviews... thanks for catching that!

sethrj committed:
Fix gdml messages not going through our logger

sethrj committed:
Use exception handler for loading GDML too

sethrj committed:
Clean up status messages

sethrj committed:
Make scoped logger thread-local

sethrj committed:
Fix logger behavior when loading geantgeoparams

Because G4GdmlParser got initialized *after* the GeantGeo, the logger
got mysteriously reset.

pcanal commented on src/celeritas/ext/GeantGeoParams.cc:
```diff
@@ -52,20 +51,17 @@ GeantGeoParams::GeantGeoParams(std::string const& filename)
+    host_ref_.world = load_geant_geometry(filename);
     loaded_gdml_ = true;
 
+    scoped_logger_ = std::make_unique<ScopedGeantLogger>();
```
```suggestion
    // Must be after the `G4GdmlParser` initialization as it (seems to) reset the loggers.
    scoped_logger_ = std::make_unique<ScopedGeantLogger>();
```

sethrj committed:
Add comment about important code ordering

pcanal reviewed (approved):
LGTM
